### PR TITLE
Adds ability to solve rigid impact problems.

### PR DESCRIPTION
**Pull request by edrumwri**
This PR is a modified version of already existing data structures for computing contact forces under the rigid contact model toward a simultaneous, inelastic rigid impact model.

This PR will form the basis of this quarter's time stepping system.

Partially and fully elastic impacts will be handled in the future.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6756)

### CODE REVIEW

**Comment by edrumwri** [comment]
+@siyuanfeng-tri for feature review please (Siyuan, this PR should look 90% identical to what you've already reviewed)

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-:-Kq_2BKrNmhrHhHTvFYC:b7tbby8)*

**Comment by siyuanfeng-tri** [comment]
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/examples/rod2d/rod2d-inl.h, line 619 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcH8Q0J_SIY7gtZa1r:-KqcH8Q17RvyY5cHm8SB:bm8bigr) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/examples/rod2d/rod2d-inl.h#L619)):*
> ```objc
>                                   std::placeholders::_1);
> 
>   // The normal and tangent spanning direction are unique.
> ```

BTW, I am not sure what this comment is supposed to mean. 

---

*[drake/multibody/rigid_contact/rigid_contact_problem_data.h, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcIs9BkqXfkPcNPs_U:-KqcIs9BkqXfkPcNPs_V:b-ajji1p) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_problem_data.h#L125)):*
> ```objc
>   /// The ℝⁿʳˣᵐ Jacobian matrix that transforms generalized velocities (m is the
>   /// dimension of generalized velocity) into velocities projected along
>   /// k vectors that span the contact tangents (used to linearize the friction
> ```

k / 2 or just r? 

---

*[drake/multibody/rigid_contact/rigid_contact_problem_data.h, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcJE1bkc6SQ0Vts5vp:-KqcJE1bkc6SQ0Vts5vq:b-mrew38) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_problem_data.h#L133)):*
> ```objc
>   MatrixX<T> F;
> 
>   /// The ℝᵐ vector v, the generalized velocity immediately before any impulsive
> ```

R^m? this is generalized v, right? 

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 167 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcLFiHuQIQ4b4EjDod:-KqcLFiHuQIQ4b4EjDoe:b-nsey9y) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L167)):*
> ```objc
>   ///        vectors should be expressed in the global frame.
>   /// @param[out] contact_impulses a non-null vector of a doublet of values,
>   ///             where the iᵗʰ triplet represents the impulsive force along
> ```

BTW ith element? 

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 177 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcLVThejJh1wTKnL0K:-KqcLVTifDY2bLQLu9d0:brw46z1) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L177)):*
> ```objc
>   ///         orthonormal.
>   /// @note If, after returning, @p contact_frames[i] is multiplied by
>   ///       @p contact_impulses[i], the result will be the contact impulses
> ```

BTW, I'd rephrase this to "after return, the ith contact impulse in the world frame is @p contact_frames[i] * @p contact_impulses[i]"

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 483 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcYmGaUzz_q6jHHd5h:-KqcYmGaUzz_q6jHHd5i:bcl0zq) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L483)):*
> ```objc
> 
>   // Construct the LCP matrix. First do the "normal contact direction" rows:
>   // N⋅M⁻¹⋅(Nᵀ - μQᵀ)  N⋅M⁻¹⋅Dᵀ  0
> ```

no -uQ^T? 

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 491 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcWYOc1-HQCo-yKZRZ:-KqcWYOc1-HQCo-yKZR_:bc35bg8) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L491)):*
> ```objc
>   const int nr = num_spanning_vectors;  //   readable construction...
>   const int nk = nr * 2;                //    of MM/qq.
>   const int num_vars = nc*2 + nk;
> ```

BTW space. 

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 498 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcWrZZa28VAlUOPcWl:-KqcWrZZa28VAlUOPcWm:bdp30i0) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L498)):*
> ```objc
>   MM->block(0, nc, nc, nr) = N * M_inv_x_FT;
>   MM->block(0, nc + nr, nc, nr) = -MM->block(0, nc, nc, nr);
>   MM->block(0, nc + nk, num_contacts, num_contacts).setZero();
> ```

BTW, why not just use nc here? 

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 521 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqca9SF_EbANDHiunf_:-Kqca9SF_EbANDHiunfa:b-lv9h8w) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L521)):*
> ```objc
> 
>   // Construct the LCP vector:
>   // N⋅M⁻¹⋅v
> ```

no M^-1? 

---

*[drake/multibody/rigid_contact/test/rigid_contact_solver_test.cc, line 136 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcQqNZCVOp5cMcbyKk:-KqcQqNZCVOp5cMcbyKl:bohqnvd) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/test/rigid_contact_solver_test.cc#L136)):*
> ```cpp
>     std::vector<double> tangent_vels;
>     rod_->GetContactPoints(*context_, &contacts);
>     rod_->GetContactPointsTangentVelocities(*context_, contacts, &tangent_vels);
> ```

do you need tan_vel?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*

**Comment by edrumwri** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/rigid_contact/rigid_contact_problem_data.h, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcIs9BkqXfkPcNPs_U:-KqciFrn6NORaBGeNTq1:b-rhu5cy) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_problem_data.h#L125)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

k / 2 or just r? 
</blockquote></details>

Done (and fixed in the acceleration problem data too).

---

*[drake/multibody/rigid_contact/rigid_contact_problem_data.h, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcJE1bkc6SQ0Vts5vp:-KqciOuqYR1yF_xAhjJv:b33pwjn) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_problem_data.h#L133)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

R^m? this is generalized v, right? 
</blockquote></details>

Right- it states in the definition of "N" that m is the dimension of the generalized velocities.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 167 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcLFiHuQIQ4b4EjDod:-KqcihIaRrIC_WB80IbE:bjrbwt4) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L167)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW ith element? 
</blockquote></details>

Copypasta- fixed here and in the acceleration-based method.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 177 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcLVThejJh1wTKnL0K:-KqcjCsAJhJbirIn6r8Q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L177)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, I'd rephrase this to "after return, the ith contact impulse in the world frame is @p contact_frames[i] * @p contact_impulses[i]"
</blockquote></details>

Done.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 483 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcYmGaUzz_q6jHHd5h:-KqcjH4ItcZjjWGqwoKs:b-xkwuoe) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L483)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

no -uQ^T? 
</blockquote></details>

Done. (Copypasta- comment was wrong)

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 491 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcWYOc1-HQCo-yKZRZ:-Kqcjsy8uj9rlBfYgBjH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L491)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW space. 
</blockquote></details>

Done.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 498 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcWrZZa28VAlUOPcWl:-KqckAFeg7xlGMdlaOE6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L498)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, why not just use nc here? 
</blockquote></details>

Done.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 521 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqca9SF_EbANDHiunf_:-KqckJctAlEFJVJHgfKH:b-mgxpmo) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_solver.h#L521)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

no M^-1? 
</blockquote></details>

More copypasta- comment was wrong.

---

*[drake/multibody/rigid_contact/test/rigid_contact_solver_test.cc, line 136 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcQqNZCVOp5cMcbyKk:-KqckN9qVT_RB578LFp1:b-vn0jov) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/test/rigid_contact_solver_test.cc#L136)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

do you need tan_vel?
</blockquote></details>

Done (forgot to remove that).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*

**Comment by edrumwri** [comment]
Review status: 4 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/rod2d/rod2d-inl.h, line 619 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcH8Q0J_SIY7gtZa1r:-Kqcn4aTM-5Ez2jiMwbz:btqo73b) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/examples/rod2d/rod2d-inl.h#L619)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, I am not sure what this comment is supposed to mean. 
</blockquote></details>

PTAL and see whether it's clear now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*

**Comment by siyuanfeng-tri** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/rigid_contact/rigid_contact_problem_data.h, line 133 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqcJE1bkc6SQ0Vts5vp:-KqcmWmukP0iowzFHeSd:b-yaxpdb) ([raw file](https://github.com/robotlocomotion/drake/blob/08c0450197f5b2df7173dbea3092c567d5d538bb/drake/multibody/rigid_contact/rigid_contact_problem_data.h#L133)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Right- it states in the definition of "N" that m is the dimension of the generalized velocities.
</blockquote></details>

i thought the superscript is n... my eyes hurt. i hate unicode. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-:-KqcnhnTEqYYZdFvkbQR:bnfp4nl)*

**Comment by edrumwri** [comment]
Wow, thanks for the superfast review, Siyuan!

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-:-KqcqfqbGfh-YB6D-h-d:b-rkcw0y)*

**Comment by edrumwri** [comment]
+@soonho-tri for feature review please (per regular rotation)

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-:-KqcqmtbQ3-CeD-9iub9:b-5blbj)*

**Comment by soonho-tri** [comment]
:lgtm: (aside minor comments) Thanks for the patience!

---

Reviewed 4 of 7 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/rod2d/rod2d-inl.h, line 616 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdbhobTkoNKGLRQ616:-KqdbhobTkoNKGLRQ617:b-dio2ol) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L616)):*
> ```objc
> 
>   // Set the inertia solver.
>   data->solve_inertia = std::bind(&Rod2D<T>::solve_inertia, this,
> ```

BTW, maybe the following is slightly more readable?
```c++
  data->solve_inertia = [this](const MatrixX<T>& m) {
    return solve_inertia(m);
  };
```

---

*[drake/examples/rod2d/rod2d-inl.h, line 632 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqde-mh5LjVhavouFRy:-Kqde-mh5LjVhavouFRz:bw02b9y) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L632)):*
> ```objc
> 
>   // Set spanning friction cone directions (set to unity, because rod is 2D).
>   data->r.resize(num_contacts);
> ```

BTW, the following line does the same task:
```c++
  data->r = std::vector<int>(num_contacts, 1);
```

---

*[drake/examples/rod2d/rod2d-inl.h, line 643 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqde5sE7R204sirqSWH:-Kqde5sE7R204sirqSWI:bbla3u0) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L643)):*
> ```objc
> 
>   // Form the tangent directions contact Jacobian (F).
>   const int nr = std::accumulate(data->r.begin(), data->r.end(), 0);
> ```

I think it's always the case that `nr == num_contacts` since rod is 2D. If you can confirm this, then we can eliminate `nr` here.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 288 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdhB03YVQYCrgGkLpE:-KqdhB03YVQYCrgGkLpF:b-2t7uph) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/multibody/rigid_contact/rigid_contact_solver.h#L288)):*
> ```objc
>                                "of elements in 'mu'");
>   }
>   const int num_spanning_vectors = std::accumulate(problem_data.r.begin(),
> ```

BTW, this line can be moved to line 298 (after `if-return`).

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 335 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdhtEfByw5AkHQ0KLF:-KqdhtEfByw5AkHQ0KLG:b9wnq2g) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/multibody/rigid_contact/rigid_contact_solver.h#L335)):*
> ```objc
>           ww.minCoeff() < -num_vars * npivots * zero_tol ||
>           abs(zz.dot(ww)) > num_vars * num_vars * npivots * zero_tol))) {
>     throw std::runtime_error("Unable to solve LCP- more regularization might "
> ```

FYI, consider the following formatting (from `clang-format-3.9`):
```c++
  if (!success ||
      (zz.size() > 0 &&
       (zz.minCoeff() < -num_vars * npivots * zero_tol ||
        ww.minCoeff() < -num_vars * npivots * zero_tol ||
        abs(zz.dot(ww)) > num_vars * num_vars * npivots * zero_tol))) {
```

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 486 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdlAYC6t0b6l73CSZB:-KqdlAYC6t0b6l73CSZC:b-4vhmi0) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/multibody/rigid_contact/rigid_contact_solver.h#L486)):*
> ```objc
>   // where D = |  F |
>   //           | -F |
>   const int nc = num_contacts;          // Alias these vars for more...
> ```

BTW, probably better to just write a single-line comment such as:
```c++
 // Alias these vars for more readable construction of MM/qq.
 ```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-:-KqdnuBWWyKuXfPUKG1X:bi6a60a)*

**Comment by edrumwri** [comment]
Review status: 5 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/rod2d/rod2d-inl.h, line 616 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdbhobTkoNKGLRQ616:-Kqe1JPoQ_XPgrNBKsfi:b-7pba0x) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L616)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, maybe the following is slightly more readable?
```c++
  data->solve_inertia = [this](const MatrixX<T>& m) {
    return solve_inertia(m);
  };
```
</blockquote></details>

Done. (Much more readable!)

---

*[drake/examples/rod2d/rod2d-inl.h, line 632 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqde-mh5LjVhavouFRy:-Kqds2HzCAbQPoH9DZSI:b2nbqo0) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L632)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, the following line does the same task:
```c++
  data->r = std::vector<int>(num_contacts, 1);
```
</blockquote></details>

BTW True, but it causes an unnecessary heap allocation. Wouldn't be the end of the world in this example, though.

---

*[drake/examples/rod2d/rod2d-inl.h, line 643 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqde5sE7R204sirqSWH:-KqdtfHGHNikJMEDHuh7:bft6qe6) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L643)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I think it's always the case that `nr == num_contacts` since rod is 2D. If you can confirm this, then we can eliminate `nr` here.
</blockquote></details>

Absolutely right, but since I expect people to use this code as a guide in their own systems, I don't want to encourage any such shortcuts. Thoughts?

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 288 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdhB03YVQYCrgGkLpE:-KqdtqWr0yb5I7ulRjXn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/multibody/rigid_contact/rigid_contact_solver.h#L288)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, this line can be moved to line 298 (after `if-return`).
</blockquote></details>

Done.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 335 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdhtEfByw5AkHQ0KLF:-KqduN_aT0E3l_9cKt53:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/multibody/rigid_contact/rigid_contact_solver.h#L335)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, consider the following formatting (from `clang-format-3.9`):
```c++
  if (!success ||
      (zz.size() > 0 &&
       (zz.minCoeff() < -num_vars * npivots * zero_tol ||
        ww.minCoeff() < -num_vars * npivots * zero_tol ||
        abs(zz.dot(ww)) > num_vars * num_vars * npivots * zero_tol))) {
```
</blockquote></details>

Done.

---

*[drake/multibody/rigid_contact/rigid_contact_solver.h, line 486 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-KqdlAYC6t0b6l73CSZB:-KqdudieCyqGgScg1PG1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/multibody/rigid_contact/rigid_contact_solver.h#L486)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, probably better to just write a single-line comment such as:
```c++
 // Alias these vars for more readable construction of MM/qq.
 ```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*

**Comment by edrumwri** [comment]
Thanks for the review, Soonho!

---

Review status: 5 of 7 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-:-Kqe2Obh5natwfoU1dW5:bwyez23)*

**Comment by soonho-tri** [comment]
Reviewed 1 of 2 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/rod2d/rod2d-inl.h, line 643 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6756#-Kqde5sE7R204sirqSWH:-Kqe4Et6D-1MCCRGP3_x:b1w3xpo) ([raw file](https://github.com/robotlocomotion/drake/blob/1e331d2d60e0f6c254879f56f988ebd6bc044b50/drake/examples/rod2d/rod2d-inl.h#L643)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Absolutely right, but since I expect people to use this code as a guide in their own systems, I don't want to encourage any such shortcuts. Thoughts?
</blockquote></details>

Got it. I'm convinced.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*

**Comment by soonho-tri** [comment]
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*

**Comment by siyuanfeng-tri** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6756)*
